### PR TITLE
[PagedAttention] Add bidirectional attention mask within image groups

### PR DESCRIPTION
**Pull request by p-wysocki**
### Details:
 - Enable bidirectional attention within image groups for Gemma3, where tokens can attend to future tokens if they describe the same image group. For example on the left there's full attention and on the right there's sliding window attention, where image is described using 5 image tokens
image

 - Implementation is based on how transformers implements this functionality, which uses a tokenizer output `token_type_ids`, which classifies tokens (image vs text).:

https://github.com/huggingface/transformers/blob/a6ef2a6f3549dd3267e8f5bafe4976a3217784bb/src/transformers/models/gemma3/modular_gemma3.py#L783-L796

 - This PR allows PagedAttention to take the graph input `token_type_ids` and calculate the bidirectional image attention, modifying causal mask behavior. The behavior is unchanged when `token_type_ids` is not given.

## Tests:
 - Tested locally on Gemma3
 - Fixes https://github.com/openvinotoolkit/openvino/issues/34198

### Questions/TODO:
- Do other tokenizers implement `token_type_ids` in a different way? Gemma3 docs don't seem exactly aligned with what the tokenizer implements.
  - **tokenizer output**: 0 if token is a text token, 1 if token is an image token
  - **documentation**: it says that it [distinguishes sentences](https://huggingface.co/docs/transformers/main/en/model_doc/gemma3#transformers.Gemma3Model.forward.token_type_ids), which I suppose is different image vs text.
- Should we keep the mask generation inside PagedAttention? SDPA has it the attention mask as an optional input, which produces causal mask by default. Perhaps we should construct attention masks outside of PagedAttention and feed them to the operator? This is especially important given the fact that we're also adding tree mask. 

### Tickets:
 - 171180

### CODE REVIEW

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp [hunk 1]
```diff
@@ -197,6 +198,15 @@ static std::shared_ptr<ov::Node> handle_baichuan2_13b_alibi(
     return res_alibi_slopes;
 }
 
+static std::shared_ptr<ov::Node> handle_gemma3_token_type_ids(
+    const std::map<std::string, std::shared_ptr<v0::Parameter>>& optional_model_wide_params) {
+    if (optional_model_wide_params.find("token_type_ids") != optional_model_wide_params.end()) {
+        auto param = optional_model_wide_params.at("token_type_ids");
+        return std::make_shared<v0::Convert>(param, ov::element::i32);
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp]: I'd suggest inserting Convert only after a check the 'token_type_ids' input is not i32. Perhaps, this is always true in real scenarios, but we'll make sure there's no unnecessary Convert in edge cases.

File: src/plugins/intel_cpu/src/nodes/paged_attn.cpp [hunk 2]
```diff
@@ -207,6 +207,11 @@ void PagedAttention::initSupportedPrimitiveDescriptors() {
                 ov::element::i32,
                 getInputShapeAtPort(PagedAttentionExecutor::ID_ADAPTIVE_RKV_DIVERSITY_BLOCK_SET_INDICES_BEGINS)));
 
+    // token_type_ids, i32, [B_token | 0] or [1, B_token]
```
Comment (wangleis) [inline src/plugins/intel_cpu/src/nodes/paged_attn.cpp:210]: As a common interface of PA, could you encode ids inputs and decode it during execution?

For example, using [3,9] to replace [0,0,0,1,1,1,1,1,1,1,0,0,0,0,0,0,0,...]

Comment (p-wysocki) [reply src/plugins/intel_cpu/src/nodes/paged_attn.cpp:210]: @rkazants @Wovchena I think we should make a final decision on this change, since it would be a significant refactor of the PR.

File: src/core/src/pass/sdpa_to_paged_attention.cpp [hunk 3]
```diff
@@ -151,6 +151,13 @@ bool ov::pass::SDPAToPagedAttention::run_on_model(const std::shared_ptr<ov::Mode
     ResultVector score_results;
     ResultVector adaptive_rkv_diversity_results;
 
+    if (auto token_type_ids_param = get_parameter(model, "token_type_ids")) {
+        token_type_ids_param->set_partial_shape(PartialShape{1, -1});
+        token_type_ids_param->set_element_type(element::i64);
+        token_type_ids_param->validate_and_infer_types();
+        optional_model_wide_params["token_type_ids"] = token_type_ids_param;
+    }
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/src/pass/sdpa_to_paged_attention.cpp:157]: This is a valid concern. Could you elaborate on that change?
Why overwrite shape and element type on existing parameter?

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp (hunk 1 above)
Comment (mitruska) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp]: Agree, also let's sync this Convert with Parameter changes form [comment ](https://github.com/openvinotoolkit/openvino/pull/34111/changes#r2864164938).
If any Reshape is needed it should be also applied in this function at graph level rather than "set_shape" on Parameter.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 4]
```diff
@@ -717,7 +763,8 @@ struct MHAHelper {
                               const ScoreAggregationInfo* score_info_ptr,
                               const PlainTensor& sinks,
                               size_t batch_in_seq = 0,
-                              const std::vector<PlainTensor>& sparse_attention_mask = {}) {
+                              const std::vector<PlainTensor>& sparse_attention_mask = {},
+                              size_t q_token_start = 0) {
```
Comment (mitruska) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:771]: Please extend the docstrings above, covering this added `q_token_start ` attribute.
(some of other attrs are already missing there, but let's keep it documented at least for this PR changes).

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 5]
```diff
@@ -460,6 +460,52 @@ struct MHAHelper {
...
+            }
+        }
+    }
+
+    // Return the adjusted ncausal that extends to the image group end for image tokens.
+    // For text tokens, returns the original ncausal unchanged.
+    size_t get_ncausal(size_t q_global_idx, size_t default_ncausal, size_t cur_kv_len) const {
+        if (!_token_type || q_global_idx >= _image_group_end.size()) {
+            return default_ncausal;
+        }
```
Comment (mitruska) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:504]: This initial check is performed for every call of `get_ncausal`, with the proposed common updates it will be called for every index calculation even if the "token_type_ids" input is not provided.
Consider optimizing it to not affect existing logic.

Perf validation in needed to ensure supported models are not affected by those PA changes.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 6]
```diff
@@ -2304,7 +2373,12 @@ struct AttentionExecutor : public PagedAttentionExecutor {
              output_emb,
              output_score,
              sparse_attention_mask,
-             output_arkv_similarity);
+             output_arkv_similarity,
+             token_type_ids);
+
+        if (token_type_ids) {
+            _helper.set_token_type(token_type_ids, subsequence_begins, past_lens);
```
[the remark this answers is not in the document]
Comment (p-wysocki) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:2387]: Tensor clearing has been implemented.

File: src/core/src/pass/sdpa_to_paged_attention.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (p-wysocki) [reply src/core/src/pass/sdpa_to_paged_attention.cpp:157]: It was a leftover from previous debugging sessions. I removed the reshape and convert.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp (hunk 5 above)
Comment (p-wysocki) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:504]: I made the job easier by providing a `false` check which should be quickly evaluated by CPU prediction logic, but one way or another, performance testing needs to be done. Is there a standard procedure for testing PA performance? cc @maxnick

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp (hunk 4 above)
Comment (p-wysocki) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:771]: Done.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 7]
```diff
@@ -1368,8 +1418,10 @@ struct MHAHelper {
 
         auto loop_softmax = [&](size_t b, size_t h, size_t pq) {
             auto cur_kv_len = static_cast<size_t>(past_lens.ptr<int32_t>()[b]) + 1;
-            auto ncausal = cur_kv_len;
-            // apply attention mask & sofmax
+            auto q_token_start = static_cast<size_t>(subsequence_begins.ptr<int32_t>()[b]);
+            auto ncausal = get_ncausal(q_token_start + pq, cur_kv_len, cur_kv_len);
```
[the remark this answers is not in the document]
Comment (p-wysocki) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:1426]: Fixed.

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp (hunk 1 above)
Comment (p-wysocki) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp]: Both done.

File: src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp [hunk 8]
```diff
@@ -602,9 +620,10 @@ ov::pass::StateManagementPattern::StateManagementPattern(
                 offset = std::make_shared<v0::Convert>(offset, element::i32);
             }
             sliding_window = std::make_shared<v1::Subtract>(v0::Constant::create(element::i32, Shape{}, {2}), offset);
-        } else if (pattern_map.count(gpt_oss_offset)) {
-            auto offset = pattern_map.at(gpt_oss_offset).get_node_shared_ptr();
-            if (pattern_map.at(gpt_oss_offset).get_partial_shape().rank() != 0) {
+        } else if (pattern_map.count(gptoss_gemma3_offset)) {
+            *is_gptoss_gemma3 = true;
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp:624]: I don't see how this uniquely identifies Gemma 3. The flag is going to be true in case of gpt-oss as well, so you'll get the token_type_ids input for it as well.

Comment (p-wysocki) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp:624]: GenAI does not generate `token_type_ids` input for gpt-oss, so it will not be given to PA in the transformation stage.

Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp:624]: Discussed offline that this is going to be improved with refactoring later on. Resolved.

Comment (mitruska) [reply src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp:624]: I can see it is currently distinguished by `handle_gemma3_token_type_ids` helper returning default Constant{0} as for other cases and the actual input will be provided only if there is `token_type_ids` in `optional_model_wide_params`.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 9]
```diff
@@ -460,6 +460,54 @@ struct MHAHelper {
...
+                        const PlainTensor& past_lens) {
+        _has_image_tokens = true;
+        _token_type = token_type;
+        auto total_tokens = static_cast<int32_t>(token_type.m_dims[0]);
+        _image_group_end.resize(total_tokens);
+
+        auto seq_count = static_cast<int32_t>(past_lens.m_dims[0]);
+        for (int32_t seq = 0; seq < seq_count; seq++) {
+            auto seq_begin = subsequence_begins.ptr<int32_t>()[seq];
+            auto seq_end = subsequence_begins.ptr<int32_t>()[seq + 1];
```
Comment (mitruska) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:482]: Is `subsequence_begins` size ensuerd at this point?

It is accessed at [seq] and [seq + 1] here, so in final loop at `subsequence_begins[seq_count]`, where `seq_count = past_lens.m_dims[0]`.
It means `subsequence_begins.size() <= (seq_count + 1)` must be ensured.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp (hunk 5 above)
Comment (mitruska) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:504]: As there were no perf regression detected (shared in the ticket), I don't put it as a blocker.

But still the same check is repeated for each loop step, while the state is not changing and depends on the provided inputs - remaining the same for the same compiled model.
Approach with function ptr could be investigated as one of possible options here to make the decision once per whole kernel logic, not at each element.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 10]
```diff
@@ -796,7 +847,7 @@ struct MHAHelper {
 
             for (size_t m = q_start; m < q_end; m++) {
                 // apply attention mask & sofmax
-                auto ncausal = (cur_kv_len - q_cnt + (m - q_start) + 1);
+                auto ncausal = get_ncausal(q_token_start + m, cur_kv_len - q_cnt + (m - q_start) + 1, cur_kv_len);
```
Comment (mitruska) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:850]: Related to my previous [comment](https://github.com/openvinotoolkit/openvino/pull/34111/changes#r2864394516)

As there were no perf regression detected at this point (shared in the ticket), I don't put it as a blocker.

But the first argument of `get_ncausal` is used only when the new `token_type_ids` input is provided, while `q_token_start + m` calculated for at each loop step even if not used.

File: src/core/tests/type_prop/paged_attention.cpp [hunk 11]
```diff
@@ -45,6 +45,8 @@ TEST(type_prop, paged_attention_static_eviction_per_block) {
     const auto adaptive_rkv_diversity_block_set_indices_begins =
         std::make_shared<op::v0::Parameter>(element::i32, PartialShape{5});
 
+    const auto token_type_ids = std::make_shared<op::v0::Parameter>(ov::element::i32, ov::Shape{0});
```
Comment (mitruska) [inline src/core/tests/type_prop/paged_attention.cpp:48]: Only existing type prop tests cases have been updated - presenting dummy token_type_ids input with empty shape. Would be good to add also tests with examples of valid input and shapes.

File: src/plugins/intel_cpu/src/nodes/paged_attn.cpp (hunk 2 above)
Comment (mitruska) [reply src/plugins/intel_cpu/src/nodes/paged_attn.cpp:210]: It has been discussed that `token_type_ids` input is already in the model, so connecting it directly to the PA (as proposed in the PR) simplify the whole logic.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp (hunk 9 above)
Comment (p-wysocki) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:482]: The shape is validated here:
https://github.com/openvinotoolkit/openvino/blob/c694fbc2b6d3193e0da33e74331bf7b289cc8231/src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp#L2163

File: src/core/tests/type_prop/paged_attention.cpp (hunk 11 above)
Comment (p-wysocki) [reply src/core/tests/type_prop/paged_attention.cpp:48]: Added new tests in https://github.com/openvinotoolkit/openvino/pull/34661

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp (hunk 10 above)
Comment (p-wysocki) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:850]: Let's discuss offline
